x/ccv/provider/keeper/grpc_query.go

```diff
@@ -49,26 +50,57 @@ func (k Keeper) QueryConsumerChains(goCtx context.Context, req *types.QueryConsu
 
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	chains := []*types.Chain{}
-	for _, chainID := range k.GetAllRegisteredConsumerIds(ctx) {
-		c, err := k.GetConsumerChain(ctx, chainID)
+	lastCID, ok := k.GetConsumerId(ctx)
+	if !ok {
+		return &types.QueryConsumerChainsResponse{}, nil
+	}
+
+	consumerIDs := []string{}
+
+	// return the consumer chains in the "Launched" phase by default
+	phase := Launched
+
+	// update phase if filter flag is  set
+	if req.FilterByPhase {
+		phase = ConsumerPhase(byte(req.Phase))
+	}
+
+	switch {
+	case phase == Launched:
+		consumerIDs = append(consumerIDs, k.GetAllRegisteredConsumerIds(ctx)...)
+	case phase == Initialized:
+		consumerIDs = append(consumerIDs, k.GetInitializedConsumers(ctx)...)
+	case phase == Registered || phase == FailedToLaunch || phase == Stopped:
+		for i := uint64(0); i <= lastCID; i++ {
```

Comment:
` <= lastCID` to `< lastCID`

Reply:
Because `k.GetConsumerId(ctx)` isn't allocated yet, right?

Reply:
Actually, it's allocated before being incremented in `FetchAndIncrementConsumerId`. So that should be `<= lastCID`, right?

Reply:
Maybe I'm missing something here but `FetchAndIncrementConsumerId` does:
```
consumerId, _ := k.GetConsumerId(ctx)
k.setConsumerId(ctx, consumerId+1)
return strconv.FormatUint(consumerId, 10)
```
so if we call `FetchAndIncrementConsumerId` we get `0` but a call to `k.GetConsumerId(ctx)` would return `1`. In this way if you have `<= cID` you end up calling `.GetConsumerPhase(ctx, "1")` that would return an error (?)

Reply:
You're right. Updated.